Snap 646: Unified Partitioner For Spark & Snappy Store

rishitesh opened the pull request:
In SnappyStore partition by column translates internally to partition by expression. This ensures column data is in de serialized form and can be passed to Snappy hashing algorithm.
Snappy hash algorithm is same as that of Spark.

Commit:
In SnappyStore partition by column translates internally to partition by exprerssion. This ensures column data is in deserialized form and can be passed to Snappy's hasing algorithm.
Snappy's hash algorithm is same as that of Spark.

Commit:
Merge branch 'snappy/master' into SNAP-STORE-646

rishitesh commented:
@suranjan @kneeraj @soubhik-c @hbhanawat  Please review

Commit:
Removed debug statements.

sumwale commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/ddl/resolver/GfxdPartitionByExpressionResolver.java:
```diff
@@ -462,7 +466,13 @@ private Serializable calcRoutingObject(DataValueDescriptor dvd,
+        return Integer.valueOf(Misc.getHashCodeSnappy(dvd));
+      }else{
+        return invokeExpressionEvaluator(dvd, null, lcc);
+      }
```
I don't think this case of general PARTITION BY (expression) should be changed to mean something else (while accepting arbitrary expressions nevertheless). We should change the normal case of PARTITION BY COLUMN () instead when isSnappyStore is true.

sumwale commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/ddl/resolver/GfxdPartitionByExpressionResolver.java:
```diff
@@ -485,7 +495,13 @@ private Serializable calcRoutingObject(DataValueDescriptor[] dvds,
+        return invokeExpressionEvaluator(null, dvds, lcc);
+      }
+
+
```
see comment above

sumwale commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/ddl/resolver/GfxdPartitionByExpressionResolver.java:
```diff
@@ -714,7 +733,7 @@ public void setPartitionColumns(String[] partCols, String[] refPartCols) {
       assert newCol != null;
       this.partColsOrigOrder[index] = newCol;
-      this.columnToIndexMap.put(newCol, Integer.valueOf(index));
+      this.columnToIndexMap.put(newCol, Integer.valueOf(index))
```
revert

sumwale commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/ddl/resolver/GfxdPartitionByExpressionResolver.java:
```diff
@@ -670,6 +686,9 @@ public boolean okForColocation(GfxdPartitionResolver rslvr) {
+        if(snappyStore){ //escaping the check for snappystore.
+          return true;
+        }
         return false;
```
see comment above

rishitesh replied:
Sumedh, I agree with you on not using expression based partitioner . I was exploring and it looked like the easiest way to implement this feature. But we can not go ahead with the de serialization overhead every time we compute hashcode(). This will defeat the purpose. I am now trying to use the column based partitioner. 

Two code paths we need to check. One is from computeHashCode() , which we need to add another method in DataValueDescriptor to give a hashcode which is similar to Spark's hashcode. This part is easy. Another is when we compute hashcode from row byte[]. Couple of DataTypes like SQLDate needs to change to serialize extra information that we need make it similar to java.util.Date() hashcode. 
SQLChar also needs some modification.

sumwale replied:
For first cut we can use the expression based partitioner. However, the above code changes bypass the expressions completely and always do the new column based partitioning. That's what my primary objection was. It should continue to use expressions with a special tag/expression for the new partitioner if required. General expression based partitioning is still useful for many cases like partitioning on month/year extracting from timestamp column but these changes will not allow that to be done (we don't allow this for column tables yet, but for row tables this is still valid).

For simplicity we can interpret PARTITION BY COLUMN when isSnappyStore=true to be the special expression I mentioned above (i.e. requiresSerializedHash=false for now) and use the deserialization approach as in your current patch, then gradually add overrides to both DVD and byte[] form for different datatypes starting from SQLChar/SQLVarchar/SQLClob (and its byte form).

Commit:
Rather than using partition by expression , for snappy we will use partition by column.
Although for row buffer table we will pass a hint to SqlParser so that it will compute UnifiedHashCode on deserialized DVD.
For actual column table we dont need this as this is partitioned by bucket. If it uses unified hash code it will result in remote puts.

Commit:
Removing snappyStore check for partition by expression

Commit:
Merge branch 'snappy/master' into SNAP-646

sumwale commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/snappy/CallbackFactoryProvider.java:
```diff
@@ -38,6 +38,16 @@ public Set createCachedBatch(BucketRegion region, UUID batchID, int bucketID) {
+
+    @Override
+    public int getHashCodeSnappy(Object[] dvds) {
+      return 0;
```
Would be better to throw unsupported exceptions here.

sumwale commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/ddl/resolver/GfxdPartitionByExpressionResolver.java:
```diff
@@ -130,6 +134,8 @@ public GfxdPartitionByExpressionResolver(ValueNode node) {
     }
   }
 
+
```
revert spaces

sumwale commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/ddl/resolver/GfxdPartitionByExpressionResolver.java:
```diff
@@ -394,13 +404,13 @@ else if (pkMap != null) {
       final String colocatedTable;
       if ((colocatedTable = distDescp.getColocateTableName()) != null) {
-        this.requiresSerializedHash = ((GfxdPartitionByExpressionResolver)
+        this.requiresSerializedHash =  ((GfxdPartitionByExpressionResolver)
```
revert

sumwale commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/ddl/resolver/GfxdPartitionByExpressionResolver.java:
```diff
@@ -394,13 +404,13 @@ else if (pkMap != null) {
       else {
         // in this case we will calculate hash from serialized value during puts
-        this.requiresSerializedHash = true;
+        this.requiresSerializedHash = customHashing ? false : true;
```
!customHashing

sumwale commented on gemfirexd/core/src/main/javacc/com/pivotal/gemfirexd/internal/impl/sql/compile/sqlgrammar.jj:
```diff
@@ -12277,8 +12277,16 @@ partitioningClause(DistributionDefinitionNode distributionNode, boolean[] repPar
+  	     distributionNode.setCustomHashing(true);
+  	  }
+  	)?
+	(
```
This will allow for HASH RANGE etc also. Lets keep this optional only for COLUMN or PARTITION BY PRIMARY KEY.

rishitesh replied:
This is a hidden keyword kind of hack. Should we be worried about its abuse ?

sumwale replied:
Perhaps we can name it like SPARKHASH or similar to make intent clear? (a non-reserved keyword)

Commit:
a) Added a method getEpoch() time in SQLTimeStamp.
b_ Made getTimeInMillis public in SQLDate.

Commit:
Merge branch 'snappy/master' into SNAP-646

Commit:
Adding Sumedh's changes.
This change disallow CLOB from being part of primary key.

Commit:
Merge branch 'snappy/master' into SNAP-646

Commit:
Reverted earlier patch to disallow complex types in primary key. Re-applied the patch with newer tests.

Commit:
Added a new keyword SPARKHASH to enable custom hashing using UnifiedPartitioner.

Commit:
Merge branch 'snappy/master' into SNAP-646
